design-proposals/golden-image-delivery-and-update-pipeline.md

```diff
@@ -0,0 +1,207 @@
+
+
+
+# API Examples
+## VM + DataVolumeTemplate + VolumeReference as source
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: DataVolume
+metadata:
+   name: my-fedora-vm-image
+spec:
+   pvc:
+     accessModes:
+     - ReadWriteOnce
+     resources:
+       requests:
+         storage: 2Gi
+   source:
+     volumeReference:
+       name: fedora
+       namespace: some-namespace
+```
+
+## VolumeImportPoller Example
+
+This example polls for image tags for the quay.io/kubevirt/fedora image and imports the latest image sorting tags based on semver. As new tags are introduced, the VolumeImportPoller will import them as new PVCs and automatically manage updating the corresponding VolumeReference.
+
+* **source** specifies where to poll from
+* **pollInterval** specifies how often to look for new imports 
+* **garbageCollectOutdated** specifies whether old PVCs should be cleaned up after a new PVC is imported.
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: VolumeImportPoller
+metadata:
+   name: fedora-poller
+spec:
+   source:
+     registry:
+       url: docker://quay.io/kubevirt/fedora
+    pollInterval: 12h
+    garbageCollectOutdated: true
+status:
+  managedVolumeReference:
+      name: fedora
+```
+
+## VolumeReference Example
+
+This example shows how a PVC is linked using the claimName field on the VolumeReference spec. The VolumeReference’s status section will indicate if the claimName is linked or dangling. 
+
+The UI can use the phase field to determine if a boot source exists for a VolumeReference or not. 
+
+* **Phase: linked** indicates that the the claimName is set and that the corresponding PVC exists and is populated
+* **Phase: dangling** indicates that the claimName points to a non-existent PVC, meaning the underlying PVC has either been deleted or is not created yet. 
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: VolumeReference
+metadata:
+   name: fedora
+spec:
+    source:
+        pvc:
+            namespace: somenamespace
+            name: fedora-32-001
+status:
+    phase: linked
+```
+
+## SSP Operator CR Example
+
+This example shows how the SSP operator could expose opting into automatic image delivery and updates for images associated with their common templates. 
+
+By default, the ssp operator would simply create dangling VolumeReferences to match the VolumeReferences used as DataVolumeTemplate sources in the common templates. Users would import their own disks which would automatically bind to the dangling VolumeReferences and make them linked to a PVC. This matches the behavior that exists today.
+
+By specifying a VolumeImportPoller on the SSP CR, one can opt-in to having VolumeReferences automatically linked to imported PVCs and have that linkage update as new versions of the corresponding image are imported.
+
+```yaml
+apiVersion: ssp.kubevirt.io/v1beta1
+kind: SSP
+metadata:
+    name: ssp-cr
+spec: 
+    commonTemplates:
```

Comment:
This means SSP (or HCO) would have to default this based on the common-templates we ship and use those images for?

Reply:
>This means SSP (or HCO) would have to default this based on the common-templates we ship and use those images for? 

I don't know exactly how the HCO will want to expose this to the admin on their API yet, but the expectation is that when the auto os disk delivery/update feature is enabled at the HCO level, that the HCO knows how to craft VolumeImportPollers on the SSP CR that align with the VolumeReferences shipped in the common templates.

So if the HCO knows the admin wants fedora and centos8 golden images auto delivered and kept up-to-date, then the HCO sets a VolumeImportPollers on the SSP cr indicating both where the images should be sourced from (the container image) and what VolumeReference it pertains to. 

SSP can in turn validate that the VolumeImportPollers actually match VolumeReferences shipped in the common templates.